storage/innobase/fil/fil0crypt.cc

```diff
@@ -3450,26 +3473,49 @@ redo_log_key *redo_log_keys::generate_and_store_new_key(THD *thd) {
   return rk;
 }
 
-redo_log_key *redo_log_keys::generate_new_key_without_storing() {
+redo_log_key *redo_log_keys::fetch_or_generate_default_key(THD *thd) {
   ut_ad(m_keys.empty());
-  Encryption::random_value(reinterpret_cast<byte *>(&m_keys[0].key));
-  return &m_keys[0];
-}
-
-bool redo_log_keys::store_used_keys() noexcept {
-  /* This is a for loop, but it really only should store a key with current
-  version 0 */
-  for (const auto &item : m_keys) {
-    if (!item.second.persisted()) {
-      ut_ad(item.first == 0);
-      if (my_key_store(PERCONA_REDO_KEY_NAME, "AES", nullptr, item.second.key,
-                       ENCRYPTION_KEY_LEN)) {
-        return false;
-      }
+  std::string default_key_name{get_key_name("", 0)}; 
+  ut_ad(strlen(server_uuid) == 0);
+  ut_ad(default_key_name.length() == strlen("percona_redo:0") && 
+        memcmp(default_key_name.c_str(), "percona_redo:0", default_key_name.length()) == 0);
+
+  char *default_redo_key_type = nullptr;
+  byte *default_rkey = nullptr;
+  size_t default_klen = 0;
+
+  if (my_key_fetch(default_key_name.c_str(), &default_redo_key_type, nullptr,
+                   reinterpret_cast<void **>(&default_rkey), &default_klen)) {
+    ib::error(ER_REDO_ENCRYPTION_CANT_FETCH_DEFAULT_KEY);
+    if (thd != nullptr) {
+      ib_senderrf(thd, IB_LOG_LEVEL_WARN, ER_REDO_ENCRYPTION_CANT_FETCH_DEFAULT_KEY);
     }
+    my_free(default_redo_key_type);
+    my_free(default_rkey);
```

Comment:
all keyrings I have written do not allocate memory on error. However, it might change if others develop new keyrings and will allocate memory even in case of error. Yes, my_free can handler nullptr :
void my_free(void *ptr) {
  my_memory_header *mh;

  if (ptr == NULL) return;

and NULL == nullptr evaluates to true.